File: backend/src/solved-ac/solved-ac.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { RateLimitError } from '../common/circuit-breaker.service';

export interface SolvedAcUser {
  handle: string;
  bio: string | null;
  organizations: SolvedAcOrganization[];
  tier: number;
  rating: number;
  ratingByProblemsSum: number;
  ratingByClass: number;
  ratingBySolvedCount: number;
  ratingByVoteCount: number;
  class: number;
  classDecoration: 'none' | 'silver' | 'gold';
  solvedCount: number;
  voteCount: number;
  exp: number;
  rank: number;
  maxStreak: number;
  prolesRank: number | null;
}

export interface SolvedAcOrganization {
  organizationId: number;
  name: string;
  type: string;
  rating: number;
  userCount: number;
  voteCount: number;
  solvedCount: number;
  color: string;
}

export interface SolvedAcTagStat {
  tag: {
    key: string;
    displayNames: Array<{ language: string; name: string; short: string }>;
  };
  solved: number;
  partial: number;
  tried: number;
}

export interface SolvedAcProblemTag {
  key: string;
  solved: number;
}

const TIER_NAMES = [
  'Unrated',
  'Bronze V', 'Bronze IV', 'Bronze III', 'Bronze II', 'Bronze I',
  'Silver V', 'Silver IV', 'Silver III', 'Silver II', 'Silver I',
  'Gold V', 'Gold IV', 'Gold III', 'Gold II', 'Gold I',
  'Platinum V', 'Platinum IV', 'Platinum III', 'Platinum II', 'Platinum I',
  'Diamond V', 'Diamond IV', 'Diamond III', 'Diamond II', 'Diamond I',
  'Ruby V', 'Ruby IV', 'Ruby III', 'Ruby II', 'Ruby I',
  'Master',
];

@Injectable()
export class SolvedAcService {
  private readonly logger = new Logger(SolvedAcService.name);
  private readonly API_BASE = 'https://solved.ac/api/v3';
  private lastRequestTime = 0;
  private readonly MIN_REQUEST_INTERVAL_MS = 200;
  private rateLimitResetTime = 0;

  constructor(private prisma: PrismaService) {}

  private async waitForRateLimit(): Promise<void> {
    const now = Date.now();

    if (this.rateLimitResetTime > now) {
      throw new RateLimitError(
        `solved.ac rate limited until ${new Date(this.rateLimitResetTime).toISOString()}`,
        this.rateLimitResetTime,
      );
    }

    const elapsed = now - this.lastRequestTime;
    if (elapsed < this.MIN_REQUEST_INTERVAL_MS) {
      await new Promise(resolve => setTimeout(resolve, this.MIN_REQUEST_INTERVAL_MS - elapsed));
    }

    this.lastRequestTime = Date.now();
  }

  private handleRateLimitResponse(response: Response): void {
    if (response.status === 429) {
      const retryAfter = response.headers.get('retry-after');
      const waitMs = retryAfter ? parseInt(retryAfter) * 1000 : 60000;
      this.rateLimitResetTime = Date.now() + waitMs;
      throw new RateLimitError(
        `solved.ac rate limit hit`,
        this.rateLimitResetTime,
      );
    }
  }

  async getUserProfile(handle: string): Promise<SolvedAcUser | null> {
    try {
      await this.waitForRateLimit();

      const response = await fetch(
        `${this.API_BASE}/user/show?handle=${encodeURIComponent(handle)}`,
      );

      if (response.status === 404) {
        return null;
      }

      if (response.status === 429) {
        this.handleRateLimitResponse(response);
      }

      if (!response.ok) {
        this.logger.warn(`solved.ac API error for ${handle}: ${response.status}`);
        return null;
      }

      const data: SolvedAcUser = await response.json();
      return data;
    } catch (error) {
      // Re-throw RateLimitError as it should be handled by circuit breaker
      if (error instanceof RateLimitError) {
        throw error;
      }
      this.logger.warn(`Failed to get solved.ac profile for ${handle}:`, error);
      return null;
    }
  }

  async getUserTagStats(handle: string): Promise<SolvedAcProblemTag[]> {
    await this.waitForRateLimit();

    const response = await fetch(
      `${this.API_BASE}/user/problem_tag_stats?handle=${encodeURIComponent(handle)}`,
    );

    if (response.status === 429) {
      this.handleRateLimitResponse(response);
    }

    if (!response.ok) {
      return [];
    }

    const data: { count: number; items: SolvedAcTagStat[] } = await response.json();

    return data.items.map((item) => ({
      key: item.tag.key,
      solved: item.solved,
    }));
  }

  getTierName(tier: number): string {
    if (tier < 0 || tier >= TIER_NAMES.length) {
      return 'Unknown';
    }
    return TIER_NAMES[tier];
  }

  extractSolvedAcHandle(bio: string | null, blog: string | null): string | null {
    const textToSearch = `${bio || ''} ${blog || ''}`;

    const patterns = [
      /solved\.ac\/profile\/(\w+)/i,
      /solved\.ac\/(@?)(\w+)/i,
      /boj[\s:]+(\w+)/i,
      /백준[\s:]+(\w+)/i,
      /baekjoon[\s:]+(\w+)/i,
      /solved\.ac[\s:]+(\w+)/i,
      /acmicpc\.net\/user\/(\w+)/i,
    ];

    for (const pattern of patterns) {
      const match = textToSearch.match(pattern);
      if (match) {
        const handle = match[2] || match[1];
        return handle.replace('@', '');
      }
    }

    return null;
  }

  async fetchAndSaveProfile(
    candidateId: string,
    handle: string,
  ): Promise<boolean> {
    const user = await this.getUserProfile(handle);

    if (!user) {
      return false;
    }

    const tagStats = await this.getUserTagStats(handle);
    const tagStatsMap: Record<string, number> = {};
    for (const tag of tagStats) {
      tagStatsMap[tag.key] = tag.solved;
    }

    await this.prisma.solvedAcProfile.upsert({
      where: { candidateId },
      update: {
        handle: user.handle,
        tier: user.tier,
        tierName: this.getTierName(user.tier),
        rating: user.rating,
        solvedCount: user.solvedCount,
        voteCount: user.voteCount,
        classLevel: user.class,
        classDecoration: user.classDecoration === 'none' ? null : user.classDecoration,
        maxStreak: user.maxStreak,
        rank: user.rank,
        tagStats: tagStatsMap,
        updatedAt: new Date(),
      },
      create: {
        candidateId,
        handle: user.handle,
        tier: user.tier,
        tierName: this.getTierName(user.tier),
        rating: user.rating,
        solvedCount: user.solvedCount,
        voteCount: user.voteCount,
        classLevel: user.class,
        classDecoration: user.classDecoration === 'none' ? null : user.classDecoration,
        maxStreak: user.maxStreak,
        rank: user.rank,
        tagStats: tagStatsMap,
      },
    });

    this.logger.log(
      `Saved solved.ac profile for ${handle}: ${this.getTierName(user.tier)} (${user.solvedCount} solved)`,
    );

    return true;
  }

  async syncCandidateSolvedAc(candidateId: string): Promise<boolean> {
    const candidate = await this.prisma.candidate.findUnique({
      where: { id: candidateId },
      select: { bio: true, blog: true, githubUsername: true },
    });

    if (!candidate) {
      return false;
    }

    let handle = this.extractSolvedAcHandle(candidate.bio, candidate.blog);

    if (!handle) {
      const user = await this.getUserProfile(candidate.githubUsername);
      if (user) {
        handle = candidate.githubUsername;
      }
    }

    if (!handle) {
      this.logger.debug(`No solved.ac handle found for candidate ${candidateId}`);
      return false;
    }

    return this.fetchAndSaveProfile(candidateId, handle);
  }

  async syncAllCandidates(options?: {
    force?: boolean;
    limit?: number;
  }): Promise<{ synced: number; failed: number; skipped: number }> {
    const { force = false, limit = 100 } = options || {};

    const whereClause = force ? {} : { solvedAcProfile: null };

    const candidates = await this.prisma.candidate.findMany({
      where: whereClause,
      select: { id: true, githubUsername: true },
      take: limit,
    });

    let synced = 0;
    let failed = 0;
    let skipped = 0;

    for (const candidate of candidates) {
      try {
        const success = await this.syncCandidateSolvedAc(candidate.id);

        if (success) {
          synced++;
          this.logger.log(`[${synced}/${candidates.length}] Synced: ${candidate.githubUsername}`);
        } else {
          skipped++;
        }
      } catch (error) {
        failed++;
        this.logger.error(`Failed to sync ${candidate.githubUsername}:`, error);
      }
    }

    this.logger.log(`Sync complete: ${synced} synced, ${skipped} skipped, ${failed} failed`);
    return { synced, failed, skipped };
  }
}

File: backend-java/src/main/java/com/peopleminer/solvedac/SolvedAcService.java
package com.peopleminer.solvedac;

import com.peopleminer.domain.entity.Candidate;
import com.peopleminer.domain.entity.SolvedAcProfile;
import com.peopleminer.domain.repository.CandidateRepository;
import com.peopleminer.domain.repository.SolvedAcProfileRepository;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import java.time.Duration;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Service
@Slf4j
public class SolvedAcService {

    private static final String API_BASE = "https://solved.ac/api/v3";
    private static final String[] TIER_NAMES = {
        "Unrated",
        "Bronze V", "Bronze IV", "Bronze III", "Bronze II", "Bronze I",
        "Silver V", "Silver IV", "Silver III", "Silver II", "Silver I",
        "Gold V", "Gold IV", "Gold III", "Gold II", "Gold I",
        "Platinum V", "Platinum IV", "Platinum III", "Platinum II", "Platinum I",
        "Diamond V", "Diamond IV", "Diamond III", "Diamond II", "Diamond I",
        "Ruby V", "Ruby IV", "Ruby III", "Ruby II", "Ruby I",
        "Master"
    };

    private static final List<Pattern> HANDLE_PATTERNS = List.of(
        Pattern.compile("solved\\.ac/profile/(\\w+)", Pattern.CASE_INSENSITIVE),
        Pattern.compile("solved\\.ac/@?(\\w+)", Pattern.CASE_INSENSITIVE),
        Pattern.compile("boj[\\s:]+(\\w+)", Pattern.CASE_INSENSITIVE),
        Pattern.compile("백준[\\s:]+(\\w+)", Pattern.CASE_INSENSITIVE),
        Pattern.compile("baekjoon[\\s:]+(\\w+)", Pattern.CASE_INSENSITIVE),
        Pattern.compile("solved\\.ac[\\s:]+(\\w+)", Pattern.CASE_INSENSITIVE),
        Pattern.compile("acmicpc\\.net/user/(\\w+)", Pattern.CASE_INSENSITIVE)
    );

    private final CandidateRepository candidateRepository;
    private final SolvedAcProfileRepository solvedAcProfileRepository;
    private final WebClient webClient;

    private long lastRequestTime = 0;
    private static final long MIN_REQUEST_INTERVAL_MS = 200;

    public SolvedAcService(
            CandidateRepository candidateRepository,
            SolvedAcProfileRepository solvedAcProfileRepository
    ) {
        this.candidateRepository = candidateRepository;
        this.solvedAcProfileRepository = solvedAcProfileRepository;
        this.webClient = WebClient.builder()
            .baseUrl(API_BASE)
            .build();
    }

    private void waitForRateLimit() throws InterruptedException {
        long now = System.currentTimeMillis();
        long elapsed = now - lastRequestTime;
        if (elapsed < MIN_REQUEST_INTERVAL_MS) {
            Thread.sleep(MIN_REQUEST_INTERVAL_MS - elapsed);
        }
        lastRequestTime = System.currentTimeMillis();
    }

    public SolvedAcUser getUserProfile(String handle) {
        try {
            waitForRateLimit();

            return webClient.get()
                .uri("/user/show?handle={handle}", handle)
                .retrieve()
                .bodyToMono(SolvedAcUser.class)
                .timeout(Duration.ofSeconds(10))
                .onErrorResume(e -> {
                    log.warn("Failed to get solved.ac profile for {}: {}", handle, e.getMessage());
                    return Mono.empty();
                })
                .block();
        } catch (Exception e) {
            log.warn("Failed to get solved.ac profile for {}: {}", handle, e.getMessage());
            return null;
        }
    }

    public String getTierName(int tier) {
        if (tier < 0 || tier >= TIER_NAMES.length) {
            return "Unknown";
        }
        return TIER_NAMES[tier];
    }

    public String extractSolvedAcHandle(String bio, String blog) {
        String textToSearch = (bio != null ? bio : "") + " " + (blog != null ? blog : "");

        for (Pattern pattern : HANDLE_PATTERNS) {
            Matcher matcher = pattern.matcher(textToSearch);
            if (matcher.find()) {
                String handle = matcher.group(1);
                return handle.replace("@", "");
            }
        }

        return null;
    }

    @Transactional
    public boolean fetchAndSaveProfile(String candidateId, String handle) {
        SolvedAcUser user = getUserProfile(handle);
        if (user == null) {
            return false;
        }

        Map<String, Integer> tagStats = new HashMap<>();
        // Tag stats would be fetched separately if needed

        Candidate candidate = candidateRepository.findById(candidateId).orElse(null);
        if (candidate == null) {
            return false;
        }

        SolvedAcProfile profile = solvedAcProfileRepository.findByCandidateId(candidateId)
            .orElse(SolvedAcProfile.builder().build());

        profile.setCandidate(candidate);
        profile.setHandle(user.getHandle());
        profile.setTier(user.getTier());
        profile.setTierName(getTierName(user.getTier()));
        profile.setRating(user.getRating());
        profile.setSolvedCount(user.getSolvedCount());
        profile.setVoteCount(user.getVoteCount());
        profile.setClassLevel(user.getSolvedAcClass());
        profile.setClassDecoration("none".equals(user.getClassDecoration()) ? null : user.getClassDecoration());
        profile.setMaxStreak(user.getMaxStreak());
        profile.setRank(user.getRank());
        profile.setTagStats(tagStats);

        solvedAcProfileRepository.save(profile);

        log.info("Saved solved.ac profile for {}: {} ({} solved)",
            handle, getTierName(user.getTier()), user.getSolvedCount());

        return true;
    }

    @Transactional
    public boolean syncCandidateSolvedAc(String candidateId) {
        Candidate candidate = candidateRepository.findById(candidateId).orElse(null);
        if (candidate == null) {
            return false;
        }

        String handle = extractSolvedAcHandle(candidate.getBio(), candidate.getBlog());

        if (handle == null) {
            SolvedAcUser user = getUserProfile(candidate.getGithubUsername());
            if (user != null) {
                handle = candidate.getGithubUsername();
            }
        }

        if (handle == null) {
            log.debug("No solved.ac handle found for candidate {}", candidateId);
            return false;
        }

        return fetchAndSaveProfile(candidateId, handle);
    }

    @Transactional
    public SyncResult syncAllCandidates(boolean force, int limit) {
        List<Candidate> candidates = force
            ? candidateRepository.findAll(PageRequest.of(0, limit)).getContent()
            : candidateRepository.findCandidatesWithoutSolvedAcProfile(PageRequest.of(0, limit));

        int synced = 0;
        int failed = 0;
        int skipped = 0;

        for (Candidate candidate : candidates) {
            try {
                boolean success = syncCandidateSolvedAc(candidate.getId());
                if (success) {
                    synced++;
                    log.info("[{}/{}] Synced: {}", synced, candidates.size(), candidate.getGithubUsername());
                } else {
                    skipped++;
                }
            } catch (Exception e) {
                failed++;
                log.error("Failed to sync {}: {}", candidate.getGithubUsername(), e.getMessage());
            }
        }

        log.info("Sync complete: {} synced, {} skipped, {} failed", synced, skipped, failed);
        return SyncResult.builder()
            .synced(synced)
            .failed(failed)
            .skipped(skipped)
            .build();
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class SolvedAcUser {
        private String handle;
        private String bio;
        private int tier;
        private int rating;
        private int solvedCount;
        private int voteCount;
        private int solvedAcClass;
        private String classDecoration;
        private int maxStreak;
        private Integer rank;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class SyncResult {
        private int synced;
        private int failed;
        private int skipped;
    }
}
